apps/nextjs-website/src/app/[...slug]/page.tsx

```diff
@@ -0,0 +1,60 @@
+import { redirect } from 'next/navigation';
+import { MenuItem, Route, PageProps } from '../types';
+
+export async function generateStaticParams() {
+  const data = await fetch('http://localhost:3000/navigationData.json');
+  const jsonData: { data: Route[] } = await data.json();
+
+  if (!jsonData.data[0]) {
+    throw new Error('No route available');
+  }
+
+  const rootRoutes = jsonData.data[0].attributes.items.data;
+
+  const allParams: { slug: string[] }[] = [];
+
+  function collectParams(routes: MenuItem[], currentPath: string[] = []) {
+    for (const route of routes) {
+      const routePath = [...currentPath, route.attributes.url];
+      const dynamicParam = [routePath.join('')];
+
+      allParams.push({ slug: dynamicParam });
+
+      if (route.attributes.children.data.length > 0) {
+        collectParams(route.attributes.children.data, routePath);
+      }
+    }
+  }
+
+  collectParams(rootRoutes);
+
+  return allParams;
+}
+
+export default async function Page({ params }: PageProps) {
+  const { slug } = params;
+  const slugExists = await isSlugValid(slug);
+
+  if (!slugExists) {
+    return redirect('/404');
+  }
```

Comment:
You can use [dynamicParams](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamicparams) instead of custom logic

Reply:
Check the docs about [dynamicParams](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamicparams) with more attention. If you set it to `false` you don't need any custom logic. The system returns 404 for dynamic segments not returned by `generateStaticParams` function.